src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Receivers.kt

```diff
@@ -140,17 +142,90 @@ open class TaskReceiver(tasksDirectory: String,
     private val arguments = listOf("projectRootAbsPath", "workingDirRelPath",
             "inputFileName", "outputFileName", "mainFont")
     private val COMPILE_COMMAND_KEY = "pandoc.compile.command"
+    private val executor = Executors.newSingleThreadExecutor()
+    private val currentTasks = ConcurrentHashMap<String, Future<Boolean>>()
 
     override fun processMessage(message: PubsubMessage): Boolean {
-        val request = CompileRequest.parseFrom(message.data)
+        val request = Request.parseFrom(message.data)
+
+        when (request.typeCase) {
+            Request.TypeCase.COMPILE -> return processCompile(request.compile)
+            Request.TypeCase.CANCEL  -> return processCancel(request.cancel)
+            else -> {
+                LOGGER.error("Request type = {} is not set properly", request.typeCase)
+            }
+        }
+
+        return true
+    }
+
+    private fun processCancel(request: CancelRequestProto): Boolean {
+        LOGGER.debug("Canceling the task with id = {}", request.taskId)
+        val status = cancelTask(request)
+        val data = CompilerFleet.Cancel.newBuilder()
+                .setTaskId(request.taskId)
+                .setStatus(status)
+                .setCpuTime(0) // TODO: put value when cancel task is fixed
+                .build()
+                .toByteString()
+        val onPublishFailureCallback = {
+            LOGGER.info("Publish cancel task ${request.taskId} response failed with code ${StatusCode.FAILURE}")
+        }
+
+        resultPublisher.publish(data, onPublishFailureCallback)
+        return true
+    }
+
+    private fun cancelTask(request: CancelRequestProto): CompilerFleet.Cancel.Status {
+        return if (!currentTasks.contains(request.taskId)) {
+            CompilerFleet.Cancel.Status.NOT_FOUND
+        } else {
+            val result = currentTasks[request.taskId]?.cancel(true)
+            currentTasks.remove(request.taskId)
+            // will canceling the thread solve the problem?
+            // another idea: two tasks maps of each processing stage
+            // 1) tasks are converting to latex at the moment
+            // 2) tasks are pushed to the texbe
+            //
+            // 1) we stop by stopping the cmd process
+            // 2) we stop by sending a cancel request to the texbe
+
+            if (result != null && result) {
+                CompilerFleet.Cancel.Status.OK
+            } else {
+                CompilerFleet.Cancel.Status.FAILED
+            }
+        }
+    }
+
+    fun isTaskDone(taskId: String): Boolean {
+        return !this.currentTasks.containsKey(taskId)
+    }
+
+    private fun processCompile(request: CompileRequest): Boolean {
         LOGGER.debug("Converting Markdown to tex: {}", request.mainFileName)
 
+        if (currentTasks.contains(request.id)) {
+            LOGGER.debug("Task with id {} is already executing", request.id)
+            return true
```

Comment:
Is it okay that this method never returns false? Maybe remove return value?

Reply:
I removed returning value.